### ISSUE TITLE
GitHub OAuth2 App Deprecation

### ISSUE THREAD

**Issue by filipecifali** [pusher/oauth2_proxy#382]
I've received an e-mail regarding some oauth2_proxies that we have installed here pointing me to this link: https://developer.github.com/changes/2019-11-05-deprecated-passwords-and-authorizations-api/

Upon checking the docs, I think we are not supporting yet the "new github way"? I don't have much experience with Go, but as far as a I've read from the github.go code it's still using the access_token there.

## Current Behavior

Authentication is working fine, but we are receiving this e-mail in different orgs and there's no date for them to end the support on these APIs, which is quite concerning. 

## Possible Solution

I'm not sure how hard is to implement the changes required by the API...

## Steps to Reproduce (for bugs)

Authenticate with GitHub as provider, wait a couple of hours and GitHub should send your organisation an e-mail saying that this will soon^TM stop working.

### Context

From what I checked, from v2.2.0 to current latest v5.0.0 the code didn't change in that regard

## Your Environment

We use oauth2_proxy for a couple of different sites, they all are under 2 organisations, but only one received the e-mail at the moment. 

- Version used:
v2.2.0 (it's upgradable, but doesn't solve the issue I think?)

**Comment by silverjam** [comment]
Upgrading should resolve this issue.  In the v2.2 code, `access_token` is provided as a query parameter to GitHub, example:
- https://github.com/pusher/oauth2_proxy/blob/v2.2/providers/github.go#L64

In version 5.0.0, they set the `Authorization` header which is the preferred (non-deprecated) method:
- https://github.com/pusher/oauth2_proxy/blob/v5.0.0/providers/github.go#L91

**Comment by JoelSpeed** [comment]
@filipecifali Have you tried upgrading? Can we close this issue?

**Comment by filipecifali** [comment]
@JoelSpeed I think so, the proxies are working fine, I was just waiting the 3 days period to see if this happens again, but I think we can close for now and if it happens again I can open a new issue, thanks!

**Comment by iain-buclaw-sociomantic** [comment]
Problem still exists in v5.0.0

**Comment by iain-buclaw-sociomantic** [comment]
Abbreviated logs that appear to be the root cause (timestamp matches that sent by Github)
```
Feb 07 08:30:53 eu-368 oauth2_proxy[22677]: [2020/02/07 08:30:53] [oauthproxy.go:855] Refreshing 1h0m1.432562119s old session cookie for Session{email:john.doe@company.co...
Feb 07 08:30:53 eu-368 oauth2_proxy[22677]: [2020/02/07 08:30:53] [internal_util.go:67] 200 GET https://api.github.com/?access_token=f8baead7fa33bc25fe2b... {"current_use...
Feb 07 08:30:53 eu-368 oauth2_proxy[22677]: [2020/02/07 08:30:53] [oauthproxy.go:855] Refreshing 1h0m1.432562119s old session cookie for Session{email:john.doe@company.co...
```

**Comment by JoelSpeed** [comment]
@iain-buclaw-sociomantic looks as though you are correct, this is fixed for the initial login, but not for refreshing.

Though we should be able to fix that by adding a `valdiateToken` to the GitHub provider which sets the appropriate header right?

**Comment by iain-buclaw-sociomantic** [comment]
I imagine so, I won't profess myself as knowledgable in either Github's API nor this project's codebase. :-)

**Comment by filipecifali** [comment]
@iain-buclaw-sociomantic @JoelSpeed this behaviour only happens if cookie-refresh is set isn't it? The default setting for cookie-expire = 168h + cookie-refresh = 0 should make the user re-login in the instance without causing the refresh code to be triggered. Thanks anyway :)

**Comment by iain-buclaw-sociomantic** [comment]
@filipecifali Unless something else has happened which would force a revalidation, then yes I think you're right.

**Comment by filipecifali** [comment]
@iain-buclaw-sociomantic then double thanks for noticing it! I wouldn't see that in the code until this happens :)

**Comment by matiasgarciaisaia** [comment]
This has been fixed by #385
